app/services/CampaignService.scala

```diff
@@ -97,4 +94,103 @@ object CampaignService {
+
+      for {
+        section <- sectionOrError
+        updatedCampaign = CampaignTransformer.updateExistingCampaign(section, campaign, user)
+        _ <- updateCampaignContent(content, updatedCampaign)
+        _ <- CampaignRepository.putCampaign(updatedCampaign)
+      } yield {
+        Logger.info(s"refreshing campaign ${updatedCampaign.name} (${updatedCampaign.id})")
+        updatedCampaign
+      }
+    }
+  }
+
+  def synchroniseCampaigns()(implicit user: User): Either[CampaignCentralApiError, List[Campaign]] = {
+
+    val sectionsToCreateOrUpdate: Seq[(CapiSection, Option[Campaign])] = {
+      val currentCampaigns = CampaignRepository
+        .getAllCampaigns()
+        .toOption
+        .getOrElse(Nil)
+        .map { campaign =>
+          campaign.pathPrefix -> campaign
+        }
+        .toMap
+
+      val sections = ContentApi.getSectionsWithPaidContentSponsorship()
+
+      sections map { section =>
+        (section, currentCampaigns.get(section.id))
+      }
+    }
+
+    val campaigns: Seq[Campaign] = sectionsToCreateOrUpdate.flatMap {
+      case (section, Some(existingCampaign)) =>
+        Some(CampaignTransformer.updateExistingCampaign(section, existingCampaign, user))
+      case (section, None) =>
+        CampaignTransformer.createDefaultCampaign(section) orElse {
+          Logger.warn(s"Could not create campaign from section: ${section.id}")
+          None
+        }
+    }
+
+    val updatedCampaignsOrError: List[Either[CampaignCentralApiError, Campaign]] = campaigns.map { campaign =>
+      val sectionId = campaign.pathPrefix
+      val content   = ContentApi.loadAllContentInSection(sectionId)
+
+      val sectionOrError = ContentApi
+        .getSection(sectionId)
+        .map(Right(_))
+        .getOrElse(Left(CampaignSectionNotFound(s"Could not find section with id: $sectionId")))
+
+      for {
+        section <- sectionOrError
+        _       <- updateCampaignContent(content, campaign)
+        _       <- CampaignRepository.putCampaign(campaign)
+      } yield {
+        Logger.info(s"refreshing campaign ${campaign.name} (${campaign.id})")
+        campaign
+      }
+
+    }.toList
+
+    updatedCampaignsOrError.sequence
+  }
+
+  private def updateCampaignContent(apiContent: List[CapiContent],
+                                    campaign: Campaign): Either[CampaignCentralApiError, Seq[ContentItem]] = {
+
+    val contentItems: List[ContentItem] = CapiContentTransformer.buildContentItems(apiContent, campaign.id)
+
+    val putContentResults: List[Either[CampaignCentralApiError, PutContentItemResult]] = for {
+      item <- contentItems
+    } yield CampaignContentRepository.putContent(item)
+
+    val results: (List[CampaignCentralApiError], List[PutContentItemResult]) =
+      putContentResults.foldRight((List[CampaignCentralApiError](), List[PutContentItemResult]())) {
+        case (itemResultOrError, (ls, rs)) => itemResultOrError.fold(l => (l :: ls, rs), r => (ls, r :: rs))
+      }
```

Comment:
If only interested in first error, could do:
```
putContentResults.collectFirst {
      case Left(e) => e
    } toLeft {
      putContentResults collect {
        case Right(result) => result
      }
    }
```

Reply:
Thanks, I just moved this code from another file but will change as much clearer.